src/main/java/io/supertokens/passwordless/Passwordless.java

```diff
@@ -0,0 +1,191 @@
+                try {
+                    passwordlessStorage.createCode(info.code);
+                    return new CreateCodeResponse(info.code.deviceIdHash, info.code.id, deviceId, userInputCode,
+                            info.linkCode, info.code.createdAt);
+                } catch (DuplicateLinkCodeHashException e) {
+                    if (gotUserInputCode) {
+                        // We only need to rethrow if the user supplied both the deviceId and the userInputCode,
+                        // because in that case the linkCodeHash will always be the same.
+                        throw e;
+                    }
+                    // It's retrieable otherwise
+                } catch (UnknownDeviceIdHash e) {
+                    throw new RestartFlowException();
+                } catch (DuplicateCodeIdException e) {
+                    // Retryable, so ignored here.
+                }
+            }
+        }
+    }
+
+    private static String generateUserInputCode() {
+        // This logic is based on the idea that we wanted to incorporate letters as well as numbers in the code.
+        // We are allowing at most 2 letters in a row, to try and avoid generating slurs or other abusive codes.
+
+        // Note: this implementation gives an equal chance to either numbers or letters, so the probability of any
+        // character is lower than the probability of a number, but the distribution is uniform inside both alphabets.
+
+        SecureRandom generator = new SecureRandom();
+        StringBuilder sb = new StringBuilder();
+        int prevAlphaCharCount = 0;
+        for (int i = 0; i < 6; ++i) {
+            if ((i < 2 || prevAlphaCharCount < 2) && generator.nextBoolean()) {
+                ++prevAlphaCharCount;
+                sb.append(getRandomAlphaChar(generator));
+            } else {
+                prevAlphaCharCount = 0;
+                sb.append(getRandomNumChar(generator));
+            }
+        }
+        return sb.toString();
+    }
+
+    public static class CreateCodeResponse {
+        public String deviceIdHash;
+        public String codeId;
+        public String deviceId;
+        public String userInputCode;
+        public String linkCode;
+        public long timeCreated;
+
+        public CreateCodeResponse(String deviceIdHash, String codeId, String deviceId, String userInputCode,
+                String linkCode, long timeCreated) {
+            this.deviceIdHash = deviceIdHash;
+            this.codeId = codeId;
+            this.deviceId = deviceId;
+            this.userInputCode = userInputCode;
+            this.linkCode = linkCode;
+            this.timeCreated = timeCreated;
+        }
+    }
+
+    private static class CreateCodeInfo {
+        public final String linkCode;
+        public final PasswordlessCode code;
+
+        public CreateCodeInfo(String id, String deviceIdHash, String linkCode, String linkCodeHash, Long createdAt) {
```

Comment:
I did think about this, but I saw no reason to make it private (but there is nothing against it either).

Reply:
Ah.. normally, the convention is that if you have an static generator function like you do, the constructor should be private. Just so that it makes it impossible to create an instance of this class without calling the exposed static function - which is what is intended.